File: 송현진/백준_JAVA/1931_회의실배정.java
import java.util.*;
import java.io.*;

public class Solution {

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();

        Node[] nodes = new Node[N];
        for (int i = 0; i < N; i++) {
            int start = sc.nextInt();
            int end = sc.nextInt();

            nodes[i] = new Node(start, end);
        }

        Arrays.sort(nodes);

        int cnt = 0;
        int prevEnd = 0;
        for (int i = 0; i < N; i++) {
            Node curNode = nodes[i];
            if (prevEnd <= curNode.start) {
                prevEnd = curNode.end;
                cnt++;
            }
        }

        System.out.println(cnt);
    }
}

class Node implements Comparable<Node> {
    int start, end;

    public Node(int start, int end) {
        this.start = start;
        this.end = end;
    }

    @Override
    public int compareTo(Node o) {
        System.out.println("this.start : " + this.start + ", o.start : " + o.start);
        return o.end == this.end ? this.start - o.start : this.end - o.end;
    }
}

File: 송현진/백준_KOTLIN/1931_회의실배정.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*


fun main() {
    var br = BufferedReader(InputStreamReader(System.`in`))
    var N = br.readLine().toInt()

    var nodes = mutableListOf<Node>()
    for(i in 0 until N){
        var st = StringTokenizer(br.readLine())
        var start = st.nextToken().toInt()
        var end = st.nextToken().toInt()

        nodes.add(Node(start,end))
    }

    //그냥 정렬하면 에러 발생 ->Node클래스 내에서 정렬 기준 설정하기
    nodes.sort()

    var cnt = 0
    var prevEnd = 0
    for(i in 0 until N){
        //현재 노드 추출
        var curNode = nodes[i]
        //이전 끝난 시각 <= 현재 노드의 시작 시간
        if(prevEnd <= curNode.start){
            //이전 끝 시각 = 현재 끝 시각
            prevEnd = curNode.end
            cnt++
        }
    }
    println(cnt)
}

//data클래스로 생성 , Comparable구현
data class Node(var start:Int, var end:Int): Comparable<Node>{

    override fun compareTo(other: Node): Int {
        //끝이 같을 경우 현재 시작 - 이전 시작 기준 정렬
        //끝이 다를 경우 현재 끝 - 이전 끝 기준 정렬
        return if(this.end==other.end)  this.start-other.start else this.end-other.end
    }
}
